File: _BENCHMARK_TAXI/code/taxi.py


# NOTA: the module pyarrow is needed to read parquet files and should be installed
import time
import pandas
import pyfixest

#
# Loading the data
#

base_taxi = pandas.read_parquet("./data/nyc_taxi.parquet")

#
# Estimations
#

#
# Base (3 FEs)
time_start = time.perf_counter()
est_base = pyfixest.feols("tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type", base_taxi)
time_base = time.perf_counter() - time_start

#
# Varying slopes
# => no known equivalent in pyfixest

#
# Multiple y
base_taxi['tip_large'] = 1 * (base_taxi["tip_amount"] >= 5)
time_start = time.perf_counter()
est_multy = pyfixest.feols("tip_amount + tip_large ~ trip_distance + passenger_count | dofw + vendor_id + payment_type", base_taxi)
time_multy = time.perf_counter() - time_start

#
# Multiple VCOVs
time_start = time.perf_counter()
est_multvcov = pyfixest.feols("tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type", base_taxi)
est_multvcov_clust = est_multvcov.vcov({"CRV1": "vendor_id"})
time_multvcov = time.perf_counter() - time_start


#
# Save
#

base_results = pandas.DataFrame({
  'method': ["pyfixest.feols" for i in range(3)],
  'type': ["base", "multiple-y", "multiple-vcov"],
  'time': [time_base, time_multy, time_multvcov]
})

base_results.to_csv("./results/results_taxi_py.csv", index = False)

File: _BENCHMARK_TAXI/code/taxi.R
#------------------------------------------------------------------------------#
# Created: 2026-01-12
# ~: Taxi benchmark for R
#------------------------------------------------------------------------------#

# needed packages
if(getOption("repos") == "@CRAN@") options(repos = "https://cran.rstudio.com")
if(!requireNamespace("arrow", quietly = TRUE)) install.packages("arrow")
if(!requireNamespace("fixest", quietly = TRUE)) install.packages("fixest")

#
# Loading the data
#

base_taxi = arrow::read_parquet("./data/nyc_taxi.parquet")

#
# Estimations
#

library(fixest)

setFixest_estimation(data = base_taxi)

#
# Base (3 FEs)
time_start = Sys.time()
est_base = feols(tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type)
time_base = Sys.time() - time_start

#
# Varying slopes
time_start = Sys.time()
est_VS = feols(tip_amount ~ passenger_count | dofw + vendor_id + payment_type[trip_distance])
time_VS = Sys.time() - time_start

#
# Multiple y
base_taxi$tip_large = +(base_taxi$tip_amount >= 5)
time_start = Sys.time()
est_multy = feols(c(tip_amount, tip_large) ~ trip_distance + passenger_count | dofw + vendor_id + payment_type)
time_multy = Sys.time() - time_start

#
# Multiple VCOVs
time_start = Sys.time()
est_multvcov = feols(tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type)
est_multvcov_clust = summary(est_multvcov, vcov = vcov_cluster("vendor_id"))
time_multvcov = Sys.time() - time_start


#
# Save
#

base_results = data.frame(
  method = "fixest::feols",
  type = c("base", "varying-slopes", "multiple-y", "multiple-vcov"),
  time = c(time_base, time_VS, time_multy, time_multvcov)
)

write.csv(base_results, "./results/results_taxi_R.csv", row.names = FALSE)







